File: 142-Linked-List-Cycle-II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        visitedMap = {}

        pos = head
        while pos is not None:
            if visitedMap.get(pos):
                return pos

            visitedMap[pos] = True
            pos = pos.next

        return None
        

File: 142-Linked-List-Cycle-II.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
const detectCycle = function (head) {
  const visitedMap = {}

  let pos = head
  while (pos) {
    if (visitedMap[pos]) {
      return pos
    }
    visitedMap[pos] = true
    pos = pos.next
  }
  return null
}
